ci.yml: centralize event-kind gating in plan; fix dependency-review failing on push to main

## Observed

Failing CI run on \`main\`: https://github.com/aidanns/agent-auth/actions/runs/25245987093 (commit \`140759b\`, push from #529).

The \`Check Security / Check Dependency Review / dependency-review\` job errored:

\`\`\`
##[error]Both a base ref and head ref must be provided, either via the \`base_ref\`/\`head_ref\` config options, \`base-ref\`/\`head-ref\` workflow action options, or by running a \`pull_request\`/\`pull_request_target\`/\`merge_group\` workflow.
\`\`\`

## Cause

\`actions/dependency-review-action@v4\` requires a base/head diff and only supports \`pull_request\`, \`pull_request_target\`, and \`merge_group\` events.

After #516 retired the legacy standalone \`dependency-review.yml\` (which only fired \`on: pull_request\`), the migrated child \`.github/workflows/check-dependency-review.yml\` is now invoked exclusively from \`ci.yml\`. \`ci.yml\` fires on:

\`\`\`yaml
on:
  pull_request: { branches: [main], types: [...] }
  push: { branches: [main] }
  merge_group:
  workflow_dispatch:
\`\`\`

So the \`check-dependency-review\` child gets called on \`push: branches: [main]\` and \`workflow_dispatch\` too — and on those events the action has no diff to scan and exits non-zero.

\`check-security.yml\`'s \`Required checks passed\` aggregator treats any non-\`success\` result (including \`failure\` and \`skipped\`) as failure, so this propagates up to \`CI / Required checks passed\` red on every merge commit.

## Impact

Every push to \`main\` (i.e. every merge) now leaves \`CI\` red on the merge commit, even when nothing is wrong. PR-time runs are unaffected — the action does have a base/head pair on \`pull_request\` events.

## Fix: centralize event-kind gating in \`plan\`

Rather than scatter \`github.event_name\` and \`github.event.pull_request.head.repo\` checks across each child, fold event-kind decisions into \`ci.yml\`'s \`plan\` job alongside the existing \`label_*\` / \`*_changed\` outputs, and thread them through \`workflow_call\` inputs to the children. Single policy point, uniform contract: children take inputs and gate steps on them.

### \`plan\` outputs to add

Three booleans, because the gates aren't identical across consumers:

- **\`event_is_pull_request\`** — \`true\` iff \`github.event_name == 'pull_request'\` (extend to \`pull_request_target\` if/when it's adopted). Used by checks that need PR-only context (changelog presence/format depends on the PR diff and PR labels).
- **\`event_has_diff_base_head\`** — \`true\` iff \`github.event_name\` ∈ {\`pull_request\`, \`pull_request_target\`, \`merge_group\`}. Used by checks that just need a base/head diff (dependency-review).
- **\`event_from_external_repo\`** — \`true\` iff this is a PR event AND \`github.event.pull_request.head.repo.full_name != github.repository\` (i.e. fork PR); \`false\` for non-PR events and same-repo PRs. Used by checks whose write-token POSTs would 403 under the read-only fork-PR \`GITHUB_TOKEN\` (dependency-submission).

All three flow through \`ci.yml\` → intermediate \`check-*\` aggregators → leaf children via \`with:\` + \`inputs:\`.

### Migration set (in scope for this issue)

- **Net-new gate**: \`.github/workflows/check-dependency-review.yml\` — declare \`inputs.event_has_diff_base_head\`, gate the \`actions/checkout\` and \`dependency-check\` steps with \`if: \${{ inputs.event_has_diff_base_head }}\`. Skipping at step level (not job level) keeps the job's result \`success\` so the aggregator stays green; the \`Required checks passed\` step rejects \`skipped\` per the comment block in \`check-security.yml\`.
- **Migrate from local check**: \`.github/workflows/check-changelog.yml\` — both \`exists\` (line 70) and \`format\` (line 134) currently inline \`if [[ \"\${{ github.event_name }}\" != \"pull_request\" ]]; then ... skip=true\`. Replace with \`inputs.event_is_pull_request\` plumbed through \`check-changelog.yml\`. The release-PR head-ref skip in those same blocks (\`HEAD_REF == release/*\`) stays local — it's content-specific to changelog rules, not event-kind policy.
- **Migrate from local check**: \`.github/workflows/check-dependency-submission.yml\` — the existing fork-PR detection step (line 36–55) computes \`IS_FORK\` from \`github.event_name == 'pull_request' && head.repo.full_name != github.repository\` and gates the checkout + snapshot-submit steps on it. Replace with \`inputs.event_from_external_repo\`: drop the \`fork-check\` step entirely, gate the remaining steps on \`if: \${{ !inputs.event_from_external_repo }}\`. The semantic stays identical — fork PRs skip, everything else (push, dispatch, merge_group, internal PR) submits.

### Explicitly NOT in scope

- \`.github/workflows/ci.yml\` line 40 — \`cancel-in-progress: \${{ github.event_name == 'pull_request' }}\` is top-level concurrency configuration on \`ci.yml\` itself, not a job step gate, and \`plan\` can't feed back to \`ci.yml\`'s own header.
- \`.github/workflows/merge-bot.yml\` and \`.github/workflows/release-bot.yml\` — top-level workflows with their own triggers; not children of \`ci.yml\`'s \`plan\` and outside the centralization surface.

## Plumbing summary

- \`ci.yml\` \`plan\` job: add the three outputs, computed from \`github.event_name\` and \`github.event.pull_request.head.repo.full_name\`.
- \`ci.yml\`: add \`with:\` blocks on the \`check-security\`, \`check-changelog\` invocations (and any future consumer).
- \`check-security.yml\`: declare \`inputs.event_has_diff_base_head\` and \`inputs.event_from_external_repo\`, forward to \`check-dependency-review\` and \`check-dependency-submission\` via \`with:\`.
- \`check-dependency-review.yml\`: declare \`inputs.event_has_diff_base_head\`, gate steps on it.
- \`check-dependency-submission.yml\`: declare \`inputs.event_from_external_repo\`, drop the local \`fork-check\` step, gate the remaining steps on it.
- \`check-changelog.yml\`: declare \`inputs.event_is_pull_request\`, replace the two inline \`github.event_name\` checks.

## Verification

After the fix lands and is squash-merged to \`main\`, the resulting \`push (main)\` run of \`CI\` should be fully green: \`check-dependency-review\` short-circuits to \`success\` (no base/head error), \`check-changelog/{exists,format}\` short-circuit to \`success\`, \`check-dependency-submission\` falls through and submits the snapshot (push has full \`GITHUB_TOKEN\`), and \`Required checks passed\` aggregates clean. Fork-PR coverage is verified by inspecting a fork PR (existing or contrived) and confirming \`check-dependency-submission\` short-circuits without 403.

--- a/.github/workflows/check-changelog.yml
+++ b/.github/workflows/check-changelog.yml
@@ -39,6 +39,19 @@ name: Check Changelog
 # inspect.
 on:
   workflow_call:
+    inputs:
+      # Centralised event-kind gate from `ci.yml`'s `plan` job
+      # (issue #530). The `exists` and `format` jobs both need PR
+      # context (the diff and labels) and short-circuit on non-PR
+      # events. The release-PR head-ref skip below is retained
+      # alongside this gate — it's a content-specific changelog rule
+      # rather than event-kind policy and stays local.
+      event_is_pull_request:
+        description: >-
+          True iff the originating event is a pull_request (the
+          changelog checks need PR diff + labels).
+        required: true
+        type: boolean
 
 permissions:
   contents: read
@@ -65,9 +78,10 @@ jobs:
         id: skip
         env:
           HEAD_REF: ${{ github.event.pull_request.head.ref }}
+          EVENT_IS_PULL_REQUEST: ${{ inputs.event_is_pull_request }}
         run: |
           set -euo pipefail
-          if [[ "${{ github.event_name }}" != "pull_request" ]]; then
+          if [[ "${EVENT_IS_PULL_REQUEST}" != "true" ]]; then
             echo "no PR context — skipping changelog presence check"
             echo "skip=true" >> "${GITHUB_OUTPUT}"
             exit 0
@@ -129,9 +143,10 @@ jobs:
         id: skip
         env:
           HEAD_REF: ${{ github.event.pull_request.head.ref }}
+          EVENT_IS_PULL_REQUEST: ${{ inputs.event_is_pull_request }}
         run: |
           set -euo pipefail
-          if [[ "${{ github.event_name }}" != "pull_request" ]]; then
+          if [[ "${EVENT_IS_PULL_REQUEST}" != "true" ]]; then
             echo "no PR context — skipping changelog format check"
             echo "skip=true" >> "${GITHUB_OUTPUT}"
             exit 0

--- a/.github/workflows/check-dependency-review.yml
+++ b/.github/workflows/check-dependency-review.yml
@@ -27,6 +27,21 @@ name: Check Dependency Review
 # each.
 on:
   workflow_call:
+    inputs:
+      # Centralised event-kind gate from `ci.yml`'s `plan` job
+      # (issue #530). `actions/dependency-review-action` only supports
+      # `pull_request`, `pull_request_target`, and `merge_group` — on
+      # `push` / `workflow_dispatch` it has no base/head pair to scan
+      # and exits non-zero. Gate the steps (not the whole job) so the
+      # job's result stays `success` and `check-security.yml`'s
+      # `Required checks passed` aggregator (which rejects `skipped`)
+      # stays green.
+      event_has_diff_base_head:
+        description: >-
+          True iff the originating event carries a base/head diff
+          (`pull_request`, `pull_request_target`, `merge_group`).
+        required: true
+        type: boolean
 
 permissions:
   # Read-only access to the Dependency Graph diff — this is the scanner input.
@@ -40,11 +55,13 @@ jobs:
     runs-on: ubuntu-latest
     timeout-minutes: 5
     steps:
-      - uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6
+      - if: ${{ inputs.event_has_diff_base_head }}
+        uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6
         with:
           persist-credentials: false
 
       - name: dependency-check
+        if: ${{ inputs.event_has_diff_base_head }}
         uses: actions/dependency-review-action@v4
         with:
           # HIGH/CRITICAL is the industry-standard gating bar and matches

--- a/.github/workflows/check-dependency-submission.yml
+++ b/.github/workflows/check-dependency-submission.yml
@@ -23,6 +23,22 @@ name: Check Dependency Submission
 # receive a read-only `GITHUB_TOKEN`, so the snapshot POST would 403.
 on:
   workflow_call:
+    inputs:
+      # Centralised fork-PR gate from `ci.yml`'s `plan` job
+      # (issue #530). Replaces the previous in-job `fork-check` step
+      # that recomputed `IS_FORK` from `github.event_name` and
+      # `head.repo.full_name`. The semantic is identical: fork PRs
+      # receive a read-only `GITHUB_TOKEN` so the snapshot POST would
+      # 403, and we short-circuit to `success` rather than
+      # `if:`-skipping the job (the parent aggregator rejects
+      # `skipped`). Push / cron-equivalent / `workflow_dispatch` /
+      # `merge_group` / internal PRs all submit normally.
+      event_from_external_repo:
+        description: >-
+          True iff this is a `pull_request` event from a fork (head
+          repo differs from the workflow repo).
+        required: true
+        type: boolean
 
 permissions:
   # Required to POST to /repos/{owner}/{repo}/dependency-graph/snapshots.
@@ -33,32 +49,11 @@ jobs:
     runs-on: ubuntu-latest
     timeout-minutes: 5
     steps:
-      - name: Skip on fork PR
-        # GITHUB_TOKEN is read-only on fork PRs, so the snapshot POST
-        # would 403. Surfacing a `success` result here (rather than
-        # `if:`-skipping the whole job) keeps the parent
-        # `check-security.yml` aggregator green — its `Required checks
-        # passed` step rejects `skipped` alongside `failure` /
-        # `cancelled` / `timed_out`. Push / cron / workflow_dispatch
-        # always run in the main repo context and fall through to the
-        # snapshot submission below.
-        id: fork-check
-        env:
-          IS_FORK: ${{ github.event_name == 'pull_request' && github.event.pull_request.head.repo.full_name != github.repository }}
-        run: |
-          set -euo pipefail
-          if [[ "${IS_FORK}" == "true" ]]; then
-            echo "skip=true" >> "${GITHUB_OUTPUT}"
-            echo "Fork PR detected — skipping snapshot submission."
-          else
-            echo "skip=false" >> "${GITHUB_OUTPUT}"
-          fi
-
-      - if: steps.fork-check.outputs.skip != 'true'
+      - if: ${{ !inputs.event_from_external_repo }}
         uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6
 
       - name: Build and submit snapshot
-        if: steps.fork-check.outputs.skip != 'true'
+        if: ${{ !inputs.event_from_external_repo }}
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
           MANIFEST_PATH: .github/tool-versions.yml

--- a/.github/workflows/check-security.yml
+++ b/.github/workflows/check-security.yml
@@ -14,6 +14,27 @@ name: Check Security
 # the aggregator's `needs:` list.
 on:
   workflow_call:
+    inputs:
+      # Centralised event-kind gating from `ci.yml`'s `plan` job
+      # (issue #530). Forwarded to leaf children whose actions
+      # require a base/head diff (`check-dependency-review`) or whose
+      # write-token POSTs would 403 on fork PRs
+      # (`check-dependency-submission`).
+      event_has_diff_base_head:
+        description: >-
+          True for events that carry a base/head diff
+          (`pull_request`, `pull_request_target`, `merge_group`) — the
+          set the GitHub `actions/dependency-review-action` supports.
+        required: true
+        type: boolean
+      event_from_external_repo:
+        description: >-
+          True for `pull_request` events whose head repo differs from
+          the workflow repo (i.e. fork PRs). Fork PRs receive a
+          read-only `GITHUB_TOKEN`, so any check that POSTs to
+          `/repos/.../dependency-graph/snapshots` etc. must short-circuit.
+        required: true
+        type: boolean
 
 permissions:
   contents: read
@@ -33,12 +54,16 @@ jobs:
   check-dependency-review:
     name: Check Dependency Review
     uses: ./.github/workflows/check-dependency-review.yml
+    with:
+      event_has_diff_base_head: ${{ inputs.event_has_diff_base_head }}
     permissions:
       contents: read
       pull-requests: write
   check-dependency-submission:
     name: Check Dependency Submission
     uses: ./.github/workflows/check-dependency-submission.yml
+    with:
+      event_from_external_repo: ${{ inputs.event_from_external_repo }}
     permissions:
       contents: write
 

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -67,8 +67,61 @@ jobs:
       docs_changed: ${{ steps.filter.outputs.docs }}
       design_changed: ${{ steps.filter.outputs.design }}
       workflows_changed: ${{ steps.filter.outputs.workflows }}
+      event_is_pull_request: ${{ steps.event.outputs.event_is_pull_request }}
+      event_has_diff_base_head: ${{ steps.event.outputs.event_has_diff_base_head }}
+      event_from_external_repo: ${{ steps.event.outputs.event_from_external_repo }}
     steps:
       - uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6
+
+      - name: Compute event-kind gating booleans
+        id: event
+        # Centralise the event-kind decisions that downstream child
+        # workflows need to gate steps on (issue #530), so individual
+        # children don't replicate `github.event_name` /
+        # `head.repo.full_name` checks. Three booleans because the
+        # downstream gates are not identical:
+        #   - `event_is_pull_request`: PR-only context (changelog
+        #     presence/format depend on the PR diff and PR labels).
+        #   - `event_has_diff_base_head`: events that carry a base/head
+        #     diff (`pull_request`, `pull_request_target`, `merge_group`)
+        #     — required by `actions/dependency-review-action`.
+        #   - `event_from_external_repo`: PR from a fork. Used by checks
+        #     whose write-token POSTs would 403 under the read-only
+        #     fork-PR `GITHUB_TOKEN` (dependency-submission).
+        # User-controlled metadata (`head.repo.full_name`) is routed
+        # through env vars rather than inlined `${{ ... }}` expressions
+        # inside `run:` so a maliciously-named fork repository cannot
+        # inject shell syntax — same convention as the
+        # `Compute label-driven gating booleans` step above.
+        env:
+          EVENT_NAME: ${{ github.event_name }}
+          HEAD_REPO_FULL_NAME: ${{ github.event.pull_request.head.repo.full_name }}
+          REPOSITORY: ${{ github.repository }}
+        run: |
+          set -euo pipefail
+          if [[ "${EVENT_NAME}" == "pull_request" ]]; then
+            event_is_pull_request=true
+          else
+            event_is_pull_request=false
+          fi
+          if [[ "${EVENT_NAME}" == "pull_request" \
+              || "${EVENT_NAME}" == "pull_request_target" \
+              || "${EVENT_NAME}" == "merge_group" ]]; then
+            event_has_diff_base_head=true
+          else
+            event_has_diff_base_head=false
+          fi
+          if [[ "${EVENT_NAME}" == "pull_request" \
+              && "${HEAD_REPO_FULL_NAME}" != "${REPOSITORY}" ]]; then
+            event_from_external_repo=true
+          else
+            event_from_external_repo=false
+          fi
+          {
+            echo "event_is_pull_request=${event_is_pull_request}"
+            echo "event_has_diff_base_head=${event_has_diff_base_head}"
+            echo "event_from_external_repo=${event_from_external_repo}"
+          } >> "${GITHUB_OUTPUT}"
 
       - name: Compute label-driven gating booleans
         id: labels
@@ -148,6 +201,9 @@ jobs:
     name: Check Security
     needs: [plan]
     uses: ./.github/workflows/check-security.yml
+    with:
+      event_has_diff_base_head: ${{ needs.plan.outputs.event_has_diff_base_head == 'true' }}
+      event_from_external_repo: ${{ needs.plan.outputs.event_from_external_repo == 'true' }}
     permissions:
       # Granted to the `check-security.yml` chain. workflow_call
       # permissions are intersected at every layer — the parent must
@@ -287,6 +343,8 @@ jobs:
     name: check-changelog
     needs: [plan]
     uses: ./.github/workflows/check-changelog.yml
+    with:
+      event_is_pull_request: ${{ needs.plan.outputs.event_is_pull_request == 'true' }}
 
   required-checks-passed:
     # Repo-wide aggregator (#440 "Decisions (locked in)"): a single

